core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala

```diff
@@ -208,3 +202,213 @@ object ScriptWitness extends Factory[ScriptWitness] {
+        val spk = ScriptPubKey.fromAsmBytes(stack(1))
+        spk
+    }
+
+  }
+
+  /** Let p = c[1:33] and let P = lift_x(int(p)) where lift_x and [:] are defined as in BIP340. Fail if this point is not on the curve.
+    */
+  def p: XOnlyPubKey = controlBlock.p
+}
+
+object TaprootScriptPath {
+
+  final val annex: Byte = 0x50
+
+  final val annexOpt: Option[Byte] = Some(annex)
+
+  final val TAPROOT_CONTROL_BASE_SIZE: Byte = 33
+  final val TAPROOT_CONTROL_NODE_SIZE: Byte = 32
+  final val TAPROOT_CONTROL_MAX_NODE_COUNT: Int = 128
+
+  final val TAPROOT_CONTROL_MAX_SIZE: Int = {
+    TAPROOT_CONTROL_BASE_SIZE + TAPROOT_CONTROL_NODE_SIZE * TAPROOT_CONTROL_MAX_NODE_COUNT
+  }
+
+  final val TAPROOT_LEAF_MASK: Byte = 0xfe.toByte
+  final val TAPROOT_LEAF_TAPSCRIPT: Byte = 0xc0.toByte
+
+  def isValid(stack: Vector[ByteVector]): Boolean = {
+    if (stack.length >= 2) {
+      val controlBlock = {
+        if (stack.headOption.map(_.head) == annexOpt) {
+          //If there are at least two witness elements, and the first byte of the last element is 0x50[4],
+          // this last element is called annex a[5] and is removed from the witness stack.
+          // The annex (or the lack of thereof) is always covered by the signature and contributes to transaction weight,
+          // but is otherwise ignored during taproot validation.
+          //see: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#script-validation-rules
+          stack(1)
+        } else {
+          stack.head
+        }
+      }
+
+      val m = controlBlock.drop(33).length / 32.0
+      if (m >= 0 && m <= 128) {
+        val pubKeyBytes = controlBlock.slice(1, 33)
+        // if not whole, we do not have correct # of bytes for control block
+        m.isWhole && SchnorrPublicKey.fromBytesOpt(pubKeyBytes).isDefined
+      } else {
+        false
+      }
+    } else {
+      false
+    }
+  }
+
+  def fromStack(stack: Vector[ByteVector]): TaprootScriptPath =
+    TaprootScriptPath(stack)
+
+  def verifyTaprootCommitment(
+      controlBlock: ControlBlock,
+      program: TaprootScriptPubKey,
+      tapLeafHash: Sha256Digest): Boolean = {
+    val internalPubKey = controlBlock.p
+    val merkleRoot = computeTaprootMerkleRoot(controlBlock, tapLeafHash)
+
+    val parity = (controlBlock.bytes.head & 1) == 1
+    program.pubKey.checkTapTweak(internal = internalPubKey,
+                                 merkleRootOpt = Some(merkleRoot),
+                                 parity = parity)
+  }
+
+  /** Computes the hash of the script that is a leaf in the tapscript tree
+    * @param leafVersion
+    * @param spk
+    * @see https://github.com/bitcoin/bitcoin/blob/37633d2f61697fc719390767aae740ece978b074/src/script/interpreter.cpp#L1828
+    * @return
+    */
+  def computeTapleafHash(leafVersion: Byte, spk: ScriptPubKey): Sha256Digest = {
+    val bytes =
+      ByteVector.fromByte(leafVersion) ++ spk.bytes
+    CryptoUtil.taggedSha256(bytes, "TapLeaf")
```

Comment:
We should precompute the tag in crypto util to speed up validation

Reply:
Can be done later

Reply:
i think i'll do it later as i would rather spend the time atm on getting the script of tapscript verification done